File: LC0437_Path_Sum_3/path_sum_3.py
# Definition for a binary tree node.
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    # Count the number of paths in the subtree rooted at root whose sum is targetSum.
    def rPathSum(self, root: TreeNode, targetSum: int) -> int:
        # base case
        if not root:
            return 0

        total = self.rPathSum(root.left, targetSum - root.val) + self.rPathSum(root.right, targetSum - root.val)
        if root.val == targetSum:
            total += 1
        return total

    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        # trivial case
        if not root:
            return 0
            
        # We must consider both when (A) the path starts at root and (B) the path does not start at root.
        return self.rPathSum(root, targetSum) + self.pathSum(root.left, targetSum) + self.pathSum(root.right, targetSum)
        

File: LC0437_Path_Sum_3/path_sum_3.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func rPathSum(root *TreeNode, targetSum int) int {
	// base case
	if root == nil {
		return 0
	}

	result := rPathSum(root.Left, targetSum-root.Val) + rPathSum(root.Right, targetSum-root.Val)
	if root.Val == targetSum {
		result += 1
	}
	return result
}

func pathSum(root *TreeNode, targetSum int) int {
	// trivial case
	if root == nil {
		return 0
	}

	// We must consider when the path contains root, and when the path does NOT contain root.
	return rPathSum(root, targetSum) + pathSum(root.Left, targetSum) + pathSum(root.Right, targetSum)
}
